### PR TITLE
fix: helmfile template fails when selector matches a chart fetched with go-getter

### PR DESCRIPTION
**Pull request by yxxhero**
helmfile template fails when selector matches a chart fetched with go-getter

fix: https://github.com/helmfile/helmfile/issues/498

```golang
func (st *HelmState) ApplyOverrides(spec *ReleaseSpec) {
	if st.OverrideKubeContext != "" && spec.KubeContext == "" {
		spec.KubeContext = st.OverrideKubeContext
	}
	if st.OverrideNamespace != "" && spec.Namespace == "" {
		spec.Namespace = st.OverrideNamespace
	}
```

for this logic. when ReleaseSpec's Namespace and KubeContext are not set. we override them. else we do nothing.

Signed-off-by: yxxhero <aiopsclub@163.com>

### CODE REVIEW

**Review by itscaro — APPROVED** [review]

File: pkg/app/run.go [hunk 1]
```diff
@@ -76,9 +76,7 @@ func (r *Run) withPreparedCharts(helmfileCommand string, opts state.ChartPrepare
 		return fmt.Errorf("%v", errs)
 	}
 
-	for i := range r.state.Releases {
-		rel := &r.state.Releases[i]
-
+	for i, rel := range r.state.GetReleasesWithOverrides() {
```
Comment (joshuasimon-taulia) [inline pkg/app/run.go:79]: ah, there it is

File: pkg/state/state.go [hunk 2]
```diff
@@ -1054,7 +1054,7 @@ func (st *HelmState) PrepareCharts(helm helmexec.Interface, dir string, concurre
 			return nil, []error{err}
 		}
 	} else {
-		selected = st.Releases
+		selected = st.GetReleasesWithOverrides()
```
Comment (joshuasimon-taulia) [inline pkg/state/state.go:1057]: just curious -- did you find this using a go debugger or just by reading the code?

**Review by joshuasimon-taulia — APPROVED** [review]

File: pkg/state/state.go (hunk 2 above)
Comment (yxxhero) [reply pkg/state/state.go:1057]: Both.

**Comment by yxxhero** [comment]
@itscaro @mumoshu I think we need to refactor the code about releaseWithOverrides logic.

**Comment by yxxhero** [comment]
@itscaro @mumoshu ping

File: pkg/app/app.go [hunk 3]
```diff
@@ -1062,7 +1062,13 @@ func (a *App) visitStatesWithSelectorsAndRemoteSupport(fileOrDir string, converg
 		}
 	}
 
-	return a.visitStates(fileOrDir, opts, f)
+	// pre-overrides HelmState
+	fHelmStatsWithOverrides := func(st *state.HelmState) (bool, []error) {
+		st.Releases = st.GetReleasesWithOverrides()
```
Comment (mumoshu) [inline pkg/app/app.go:1067]: How does this resolve the issue? 🤔
The error message looked as if it was saying that the chart name that looks like a URL isn't valid. And this GetReleasesWithOverrides doesn't touch the chart name of the problematic release, right?

**Comment by yxxhero** [comment]
@mumoshu 
let me try to explain:


the key is the top key: `namespace`
```yaml
namespace: istio-system
repositories:
....
```

when using selector feature:
https://github.com/helmfile/helmfile/blob/8f59a1c18a1a1ff020aeefed3815407bd46d5067/pkg/state/state.go#L1044-L1058

it will call `GetReleasesWithOverrides` func, set top namespace to `releaseSpec` namespace. and return `map[PrepareChartKey]string`

PrepareChartKey:
```golang
type PrepareChartKey struct {
	Namespace, Name, KubeContext string
}
```
PrepareChartKey contains namespace info. but when setting chartPath, it uses origin `state.Releases`, not contain namespace info. So the chartPath can't be set correctly. see:
https://github.com/helmfile/helmfile/blob/8f59a1c18a1a1ff020aeefed3815407bd46d5067/pkg/app/run.go#L79-L93

```
for i := range r.state.Releases 
```

chartPath match logic:

```golang
	for i := range r.state.Releases {
		rel := &r.state.Releases[i]

		key := state.PrepareChartKey{
			Name:        rel.Name,
			Namespace:   rel.Namespace,  // see this: rel.Namespace will be empty. not top namespace value. So it will not match.
			KubeContext: rel.KubeContext,
		}
		if chart := releaseToChart[key]; chart != rel.Chart {
			// In this case we assume that the chart is downloaded and modified by Helmfile and chartify.
			// So we take note of the local filesystem path to the modified version of the chart
			// and use it later via the Release.ChartPathOrName() func.
			rel.ChartPath = chart
		}
	}
```

@mumoshu my English is no good. So I hope that my explanation is ok.

**Comment by yxxhero** [comment]
@mumoshu this is a very old bug that already exists so a long time.

**Comment by yxxhero** [comment]
@mumoshu kubcontext also has same bug.

File: pkg/app/app.go (hunk 3 above)
Comment (yxxhero) [reply pkg/app/app.go:1067]: pre handle `st.Releases` by `GetReleasesWithOverrides` func. So top namespace will be set correctly. GetReleasesWithOverrides can be called many times, it is idempotence for new code.

**Comment by yxxhero** [comment]
@mumoshu if have more question, please free to ping me.

**Comment by yxxhero** [comment]
@xiaomudk ping

**Comment by mumoshu** [comment]
@yxxhero Thanks for your explanation!

Can we now treat every occurrence of `state.Releases` at a read position as a bug, and all those should be modified to use `GetReleasesWithOverrides` instead, similar to you've done in this pull request?

Then, how about doing the overriding on unmarshal time, so that we can make GetReleasesWithOverrides a private function that can't be called outside of the pkg. That way, we'll never introduce any bug like this one. WDYT?

**Comment by yxxhero** [comment]
@mumoshu makes sense. I will update the code.

**Comment by yxxhero** [comment]
I looked at the logic carefully again, and to generate a complete `helmstate`, I need to go through multiple yaml unmarshals, including the loading of sub helmfiles, and then perform logical processing after merging. The current modification is really the processing after the generation of helmstate, in the subcommand Before processing, call GetReleasesWithOverrides once.
Other modification methods currently see the changes are relatively large.


@mumoshu

**Comment by yxxhero** [comment]
@mumoshu I tried to implement it.

**Comment by yxxhero** [comment]
```golang
	if ld.namespace != "" {
		if st.OverrideNamespace != "" {
			return nil, errors.New("err: Cannot use option --namespace and set attribute namespace.")
		}
		st.OverrideNamespace = ld.namespace
	}

	if ld.chart != "" {
		if st.OverrideChart != "" {
			return nil, errors.New("err: Cannot use option --chart and set attribute chart.")
		}
		st.OverrideChart = ld.chart
	}
```

After the helmstate is generated, there is another namespace coverage logic. There are too many changes. It needs to be well designed to ensure that there are no errors. @mumoshu

File: pkg/app/app_test.go [hunk 4]
```diff
@@ -1502,7 +1502,7 @@ releases:
 		SetFilter(true),
 	)
 
-	expected := "in ./helmfile.yaml: duplicate release \"foo\" found in namespace \"foo\" in kubecontext \"default\": there were 2 releases named \"foo\" matching specified selector"
+	expected := "in ./helmfile.yaml: duplicate release \"foo\" found in namespace \"foo\" in kubecontext \"foo\": there were 2 releases named \"foo\" matching specified selector"
```
Comment (mumoshu) [inline pkg/app/app_test.go:1505]: Apparently, you found a bug in the implementation and the test case here. Good job!

Comment (yxxhero) [reply pkg/app/app_test.go:1505]: @mumoshu Thanks very much.

File: pkg/app/app_list_test.go [hunk 5]
```diff
@@ -310,7 +310,7 @@ releases:
 		assert.Nil(t, err)
 	})
 
-	expected := `[{"name":"myrelease1","namespace":"","enabled":true,"installed":false,"labels":"id:myrelease1","chart":"mychart1","version":""},{"name":"myrelease2","namespace":"","enabled":false,"installed":true,"labels":"","chart":"mychart1","version":""},{"name":"myrelease3","namespace":"","enabled":true,"installed":true,"labels":"","chart":"mychart1","version":""},{"name":"myrelease4","namespace":"","enabled":true,"installed":true,"labels":"id:myrelease1","chart":"mychart1","version":""}]
+	expected := `[{"name":"myrelease1","namespace":"testNamespace","enabled":true,"installed":false,"labels":"id:myrelease1","chart":"mychart1","version":""},{"name":"myrelease2","namespace":"testNamespace","enabled":false,"installed":true,"labels":"","chart":"mychart1","version":""},{"name":"myrelease3","namespace":"testNamespace","enabled":true,"installed":true,"labels":"","chart":"mychart1","version":""},{"name":"myrelease4","namespace":"testNamespace","enabled":true,"installed":true,"labels":"id:myrelease1","chart":"mychart1","version":""}]
```
Comment (mumoshu) [inline pkg/app/app_list_test.go:313]: So you fixed not only helmfile-template but also helmfile-list to respect the override namespace, right? Good job!

Comment (yxxhero) [reply pkg/app/app_list_test.go:313]: yes. All sub-commands had been fixed.

File: pkg/app/app_test.go (hunk 4 above)
Comment (mumoshu) [reply pkg/app/app_test.go:1505]: Well I read code twice and this wasn't actually a bug. `OverrideKubeContext` is meant for "overriding" kubecontext that is already configured in the helmfile.yaml releases[].kubeContex. Same for `OverrideNamespace`. You made OverrideKubeContext actually a "default kubecontext" in this pull request, which isn't how the field is supposed to work.
That said, I think you need to revert changes on `ApplyOverrides`.

File: pkg/state/state.go [hunk 6]
```diff
@@ -372,10 +372,10 @@ var DefaultFetchOutputDirTemplate = path.Join(
 )
 
 func (st *HelmState) ApplyOverrides(spec *ReleaseSpec) {
-	if st.OverrideKubeContext != "" {
+	if st.OverrideKubeContext != "" && spec.KubeContext == "" {
```
Comment (mumoshu) [inline pkg/state/state.go:375]: We might need to revert this change as per https://github.com/helmfile/helmfile/pull/499/files#r1020827982

File: pkg/state/state.go [hunk 7]
```diff
@@ -372,10 +372,10 @@ var DefaultFetchOutputDirTemplate = path.Join(
...
 		spec.KubeContext = st.OverrideKubeContext
 	}
-	if st.OverrideNamespace != "" {
+	if st.OverrideNamespace != "" && spec.Namespace == "" {
```
Comment (mumoshu) [inline pkg/state/state.go:378]: We might need to revert this change as per https://github.com/helmfile/helmfile/pull/499/files#r1020827982

**Comment by yxxhero** [comment]
@mumoshu done and ping

File: test/e2e/template/helmfile/snapshot_test.go [hunk 8]
```diff
@@ -172,6 +182,15 @@ func TestHelmfileTemplateWithBuildCommand(t *testing.T) {
 
 			gotStr := string(got)
 			gotStr = strings.ReplaceAll(gotStr, fmt.Sprintf("chart=%s", wd), "chart=$WD")
+
+			// Replace go-getter path with $GoGetterPath
+			if config.GoGetterMode.Enabled {
+				switch config.GoGetterMode.Protocol {
+				case "git":
+					gotStr = chartGitFullPathRegex.ReplaceAllString(gotStr, `chart=$$GoGetterPath`)
```
Comment (mumoshu) [inline test/e2e/template/helmfile/snapshot_test.go:190]: @yxxhero Can we just extract this out of the two conditional blocks, and remove GoGetterMode altogether?
This replace-all operation seems to have no effect on output that doesn't involve any go-getter charts. Then, executing this unconditionally shouldn't break any non-go-getter tests, right?

Comment (yxxhero) [reply test/e2e/template/helmfile/snapshot_test.go:190]: @mumoshu I got.

**Review by mumoshu — APPROVED** [review]
LGTM! Thanks a lot for your patience and awesome work @yxxhero ☺️

**Comment by joshuasimon-taulia** [comment]
latest release is confirmed working. thanks all!
```
helmfile  -l name=acme-jx template

Building dependency release=acme-jx, chart=/Users/joshua.simon/Library/Caches/helmfile/istio-system/acme-jx/https_github_com_joshuasimon-taulia_acme_git.ref=master/charts/acme
Templating release=acme-jx, chart=/Users/joshua.simon/Library/Caches/helmfile/istio-system/acme-jx/https_github_com_joshuasimon-taulia_acme_git.ref=master/charts/acme

# Source: acme/templates/cert-manager-prod-certificate.yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
...
```
